src/main/java/com/conveyal/gtfs/GraphQLController.java

```diff
@@ -49,33 +55,28 @@ public static Object post (Request req, Response response) {
             halt(400, "Malformed JSON");
         }
         // FIXME converting String to JSON nodes and back to string, then re-parsing to Map.
-        String vars = node.get("variables").asText();
+        JsonNode vars = node.get("variables");
         String query = node.get("query").asText();
         return doQuery(vars, query, response);
     }
 
-    private static Object doQuery (String varsJson, String queryJson, Response response) {
+
+    private static Object doQuery (JsonNode varsJson, String queryJson, Response response) {
         long startTime = System.currentTimeMillis();
         if (varsJson == null && queryJson == null) {
             return GTFSGraphQL.getGraphQl().execute(IntrospectionQuery.INTROSPECTION_QUERY).getData();
         }
-        try {
-            Map<String, Object> variables = mapper.readValue(varsJson, new TypeReference<Map<String, Object>>(){});
-            ExecutionResult result = GTFSGraphQL.getGraphQl().execute(queryJson, null, null, variables);
-            List<GraphQLError> errs = result.getErrors();
-            if (!errs.isEmpty()) {
-                response.status(400);
-                return errs;
-            } else {
-                long endTime = System.currentTimeMillis();
-                LOG.info("Query took {} msec", endTime - startTime);
-                return result.getData();
-            }
-        } catch (IOException e) {
-            LOG.warn("Error processing variable JSON", e);
-            halt(404, "Malformed JSON");
+        Map<String, Object> variables = mapper.convertValue(varsJson, Map.class);
+        ExecutionResult result = GTFSGraphQL.getGraphQl().execute(queryJson, null, null, variables);
+        List<GraphQLError> errs = result.getErrors();
+        if (!errs.isEmpty()) {
+            response.status(400);
+            return result.toSpecification();
```

Comment:
The `if` and `else` clauses both contain the same final return statement, so should be factored out. Other than setting the response status to 400 the rest of the statements in both branches of this conditional are always useful. So I'd prefer to just do `if (!errs.isEmpty()) response.status(400);` and then fall through to always run what is currently in the else block.